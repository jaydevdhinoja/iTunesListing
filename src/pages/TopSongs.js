import React, { useState } from "react";

//Global Context
import { useGlobalListingContext } from "../ContextProvider"

// Components
import Grid from "../components/Grid";
import Card from "../components/Card";
import LoadMoreBtn from "../components/elements/LoadMoreBtn";
import Spinner from "../components/elements/Spinner/Spinner";

const TopSongs = () => {
  const filterItems = 20;
  const {songs, loading, error, favourites, addToFavourite} =  useGlobalListingContext()
  const [search, setSearch] = useState("");
  const [visible, setVisible] = useState(filterItems);

  const filteredSongs = songs.filter(song => {
    return (
      song["im:name"].label.toLowerCase().indexOf(search.toLowerCase()) !== -1 ||
      song["im:artist"].label.toLowerCase().indexOf(search.toLowerCase()) !== -1
    );
  });

  const loadMoreAlbums = () => {
    setVisible(prev => prev + filterItems);
  };

  if (error) return <div>Something went wrong...</div>;

  return (
    <>
      <Grid header="Top Songs" callback={s => setSearch(s)}>
        {filteredSongs.slice(0, visible).map((song, i) => (
          <Card 
            key={song.id.attributes["im:id"]} 
            item={song} 
            index={i} 
            callback={addToFavourite}
            itemType={'songs'} 
            favourite={favourites['songs'].filter(item => item === song.id.attributes["im:id"]).length > 0}
            />
        ))}
      </Grid>
      {loading && <Spinner />}

      {visible < filteredSongs.length && !loading && (
        <LoadMoreBtn text={`Load more songs`} callback={loadMoreAlbums} />
      )}
    </>
  );
};

export default TopSongs;
