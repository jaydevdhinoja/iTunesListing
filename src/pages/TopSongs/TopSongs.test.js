import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { renderWithContext, InitialState } from '../../tests/common';
import TopSongs from '.';
import { cleanup } from '@testing-library/react';

afterEach(cleanup);

describe('Page Listing', () => {
	test('renders component', () => {
		const { container } = renderWithContext(<TopSongs />, { value: InitialState });

		expect(container).toMatchSnapshot();
	});

	test('renders component with no songs', () => {
		const { queryAllByText } = renderWithContext(<TopSongs />, {
			value: { ...InitialState, songs: [] }
		});

		expect(queryAllByText('Tim McGraw & Tyler Hubbard')).toHaveLength(0);
	});

	test('renders top songs', () => {
		const { getByText } = renderWithContext(<TopSongs />, { value: InitialState });
		expect(getByText('Tim McGraw & Tyler Hubbard')).toBeInTheDocument();
	});
});
