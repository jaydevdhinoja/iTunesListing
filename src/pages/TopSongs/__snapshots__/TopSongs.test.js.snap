// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Listing renders component 1`] = `
<div>
  <section
    class="albums-grid container"
  >
    <div
      class="albums-grid__headline section-headline"
    >
      <h2>
        Top Songs
      </h2>
      <div
        class="albums-grid__headline-search"
      >
        <input
          class="search-input"
          placeholder="Search"
          type="text"
          value=""
        />
        <img
          alt="Search"
          src="search-solid.svg"
        />
      </div>
    </div>
    <ul
      class="albums-grid__list"
    >
      <li
        class="item-card"
      >
        <a
          href="/item/1545178423"
        >
          <div
            class="item-card__mobile-link"
          />
        </a>
        <div
          class="item-card__thumbnail"
          style="background-image: url(undefined);"
        >
          <img
            alt="Heart Icon"
            class="favourite-icon"
            src="heart-solid.svg"
          />
        </div>
        <h3
          class="item-card__title"
        >
          Undivided
        </h3>
        <p
          class="item-card__artist"
        >
          Tim McGraw & Tyler Hubbard
        </p>
        <div
          class="item-card__overlay"
        />
        <p
          class="item-card__overlay-number"
        >
          1
        </p>
        <span
          class="item-card__overlay-shadow"
        />
      </li>
    </ul>
  </section>
</div>
`;
